File: SpringEcomAI/src/main/java/com/telusko/service/OrderService.java
package com.telusko.service;

import com.telusko.model.Order;
import com.telusko.model.OrderItem;
import com.telusko.model.Product;
import com.telusko.model.dto.OrderItemRequest;
import com.telusko.model.dto.OrderItemResponse;
import com.telusko.model.dto.OrderRequest;
import com.telusko.model.dto.OrderResponse;
import com.telusko.repo.OrderRepo;
import com.telusko.repo.ProductRepo;
import jakarta.transaction.Transactional;
import org.springframework.ai.document.Document;
import org.springframework.ai.vectorstore.VectorStore;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class OrderService {

    @Autowired
    private OrderRepo orderRepo;

    @Autowired
    private VectorStore vectorStore;

    @Autowired
    private ProductRepo productRepo;

    public OrderResponse placeOrder(OrderRequest request) {
        Order order = new Order();

        String uniqueOrderId = "ORD" + UUID.randomUUID().toString().replace("-", "").substring(0, 10).toUpperCase();
        order.setOrderId(uniqueOrderId);
        order.setCustomerName(request.customerName());
        order.setEmail(request.email());
        order.setOrderDate(LocalDate.now());
        order.setStatus("PLACED");

        List<OrderItem> orderItems = new ArrayList<>();

        for (OrderItemRequest itemReq : request.items()) {
            Product product = productRepo.findById(itemReq.productId())
                    .orElseThrow(() -> new RuntimeException("Product not found"));

            if (product.getStockQuantity() < itemReq.quantity()) {
                throw new RuntimeException("Insufficient stock for product: " + product.getName());
            }

            product.setStockQuantity(product.getStockQuantity() - itemReq.quantity());

            productRepo.save(product);

            String filter = String.format("productId == '%s'", String.valueOf(product.getId()));
            vectorStore.delete(filter);

            String updatedContent = String.format("""
                             Product Name: %s
                             Description: %s
                             Brand: %s
                             Category: %s
                             Price: %.2f
                             Release Date: %s
                             Available: %s
                             Stock: %d
                            """,
                    product.getName(),
                    product.getDescription(),
                    product.getBrand(),
                    product.getCategory(),
                    product.getPrice(),
                    product.getReleaseDate(),
                    product.isProductAvailable(),
                    product.getStockQuantity()
            );

            Document updatedDoc = new Document(
                    UUID.randomUUID().toString(),
                    updatedContent,
                    Map.of("productId", String.valueOf(product.getId()))
            );

            vectorStore.add(List.of(updatedDoc));


            OrderItem item = OrderItem.builder()
                    .product(product)
                    .quantity(itemReq.quantity())
                    .totalPrice(product.getPrice().multiply(BigDecimal.valueOf(itemReq.quantity())))
                    .order(order)
                    .build();

            orderItems.add(item);
        }

        order.setItems(orderItems);
        Order savedOrder = orderRepo.save(order);

        StringBuilder contentToEmbed = new StringBuilder();
        contentToEmbed.append("Order Summary:\n");
        contentToEmbed.append("Order ID: ").append(savedOrder.getOrderId()).append("\n");
        contentToEmbed.append("Customer: ").append(savedOrder.getCustomerName()).append("\n");
        contentToEmbed.append("Email: ").append(savedOrder.getEmail()).append("\n");
        contentToEmbed.append("Date: ").append(savedOrder.getOrderDate()).append("\n");
        contentToEmbed.append("Status: ").append(savedOrder.getStatus()).append("\n");
        contentToEmbed.append("Products:\n");

        for (OrderItem item : savedOrder.getItems()) {
            contentToEmbed.append("- ").append(item.getProduct().getName())
                    .append(" x ").append(item.getQuantity())
                    .append(" = ₹").append(item.getTotalPrice()).append("\n");
        }

        Document document = new Document(
                UUID.randomUUID().toString(),
                contentToEmbed.toString(),
                Map.of("orderId", savedOrder.getOrderId())
        );

        vectorStore.add(List.of(document));

        List<OrderItemResponse> itemResponses = savedOrder.getItems().stream()
                .map(i -> new OrderItemResponse(
                        i.getProduct().getName(),
                        i.getQuantity(),
                        i.getTotalPrice()))
                .collect(Collectors.toList());

        return new OrderResponse(
                savedOrder.getOrderId(),
                savedOrder.getCustomerName(),
                savedOrder.getEmail(),
                savedOrder.getStatus(),
                savedOrder.getOrderDate(),
                itemResponses
        );
    }


    @Transactional
    public List<OrderResponse> getAllOrderResponses() {
        List<OrderResponse> orders = new ArrayList<>();

        orderRepo.findAll().forEach(order -> {
            List<OrderItemResponse> itemResponses = order.getItems().stream()
                    .map(i -> new OrderItemResponse(
                            i.getProduct().getName(),
                            i.getQuantity(),
                            i.getTotalPrice()))
                    .toList();

            orders.add(new OrderResponse(
                    order.getOrderId(),
                    order.getCustomerName(),
                    order.getEmail(),
                    order.getStatus(),
                    order.getOrderDate(),
                    itemResponses));
        });

        return orders;
    }

}

File: SpringEcomAI-LangChain/services/order_service.py
import uuid
from datetime import date
from decimal import Decimal

from sqlalchemy.orm import Session
from langchain_core.documents import Document

from models import Order, OrderItem, Product
from schemas import OrderRequest, OrderResponse, OrderItemResponse
from config.vector_store import vector_store


def _delete_from_vector_store_by_product(product_id: int):
    collection = vector_store._collection
    results = collection.get(where={"productId": str(product_id)})
    if results["ids"]:
        vector_store.delete(ids=results["ids"])


def place_order(db: Session, request: OrderRequest) -> OrderResponse:
    order_id = "ORD" + uuid.uuid4().hex[:10].upper()

    order = Order(
        order_id=order_id,
        customer_name=request.customer_name,
        email=request.email,
        order_date=date.today(),
        status="PLACED"
    )
    db.add(order)
    db.flush()

    created_items = []

    for item_req in request.items:
        product = db.query(Product).filter(Product.id == item_req.product_id).first()
        if not product:
            raise ValueError(f"Product not found: {item_req.product_id}")

        if product.stock_quantity < item_req.quantity:
            raise ValueError(f"Insufficient stock for: {product.name}")

        product.stock_quantity -= item_req.quantity

        _delete_from_vector_store_by_product(product.id)
        updated_doc = Document(
            page_content=f"""
Product Name: {product.name}
Description: {product.description}
Brand: {product.brand}
Category: {product.category}
Price: {float(product.price):.2f}
Available: {product.product_available}
Stock: {product.stock_quantity}
""".strip(),
            metadata={"productId": str(product.id)}
        )
        vector_store.add_documents([updated_doc], ids=[str(uuid.uuid4())])

        total = product.price * Decimal(item_req.quantity)
        order_item = OrderItem(
            product_id=product.id,
            quantity=item_req.quantity,
            total_price=total,
            order_fk=order.id
        )
        db.add(order_item)
        created_items.append((product, order_item))

    db.commit()
    db.refresh(order)

    # Index order in vector store
    content = (
        f"Order ID: {order.order_id}\n"
        f"Customer: {order.customer_name}\n"
        f"Email: {order.email}\n"
        f"Status: {order.status}\nProducts:\n"
    )
    for product, item in created_items:
        content += f"- {product.name} x {item.quantity} = {item.total_price}\n"

    order_doc = Document(page_content=content, metadata={"orderId": order.order_id})
    vector_store.add_documents([order_doc], ids=[str(uuid.uuid4())])

    item_responses = [
        OrderItemResponse(
            product_name=product.name,
            quantity=item.quantity,
            total_price=item.total_price
        )
        for product, item in created_items
    ]

    return OrderResponse(
        order_id=order.order_id,
        customer_name=order.customer_name,
        email=order.email,
        status=order.status,
        order_date=order.order_date,
        items=item_responses
    )


def get_all_orders(db: Session):
    orders = db.query(Order).all()
    result = []

    for order in orders:
        item_responses = []
        for item in order.items:
            product = db.query(Product).filter(Product.id == item.product_id).first()
            item_responses.append(OrderItemResponse(
                product_name=product.name if product else f"Deleted Product #{item.product_id}",
                quantity=item.quantity,
                total_price=item.total_price
            ))

        result.append(OrderResponse(
            order_id=order.order_id,
            customer_name=order.customer_name,
            email=order.email,
            status=order.status,
            order_date=order.order_date,
            items=item_responses
        ))

    return result
